pkg/registry/singleton.go

```diff
@@ -1,22 +1,244 @@
+	DeleteRegistryCommand       = "if docker inspect %s 2>/dev/null;then docker rm -f %[1]s;fi && ((! nerdctl ps -a 2>/dev/null |grep %[1]s) || (nerdctl stop %[1]s && nerdctl rmi -f %[1]s))"
 )
 
 type localSingletonConfigurator struct {
-	LocalRegistry
-	ContainerRuntimeInfo containerruntime.Info
+	rootfs string
+	*LocalRegistry
+
+	configFileGenerator          ConfigFileGenerator
+	containerRuntimeConfigurator containerruntime.Configurator
+	infraDriver                  infradriver.InfraDriver
+	containerRuntimeInfo         containerruntime.Info
 }
 
+// Reconcile local private registry by rootfs scripts.
 func (c *localSingletonConfigurator) Reconcile() (Driver, error) {
+	if err := c.genTLSCerts(); err != nil {
+		return nil, err
+	}
+
+	if err := c.genBasicAuth(); err != nil {
+		return nil, err
+	}
+
+	if err := c.initRegistry(); err != nil {
+		return nil, err
+	}
+
+	if err := c.configHostsFile(); err != nil {
+		return nil, err
+	}
+
+	if err := c.configKubeletAuthInfo(); err != nil {
+		return nil, err
+	}
+
+	if err := c.configRegistryCert(); err != nil {
+		return nil, err
+	}
+
+	config := containerruntime.DaemonConfig{
+		Endpoint: c.Domain + ":" + strconv.Itoa(c.Port),
+	}
+
+	if err := c.containerRuntimeConfigurator.ConfigDaemonService(config); err != nil {
+		return nil, err
+	}
+
+	return NewLocalRegistryDriver(c.DataDir, c.infraDriver), nil
+}
+
+func (c *localSingletonConfigurator) genTLSCerts() error {
+	// if deploy registry as InsecureMode ,skip to gen cert.
+	if c.InsecureMode {
+		return nil
+	}
+
+	DNSNames := []string{c.Domain}
+	DNSNames = append(DNSNames, c.Cert.SubjectAltName.IPs...)
+	DNSNames = append(DNSNames, c.Cert.SubjectAltName.DNSNames...)
+
+	regCertConfig := cert.CertificateDescriptor{
+		CommonName:   "registry-ca",
+		DNSNames:     DNSNames,
+		Organization: nil,
+		Year:         100,
+		AltNames:     cert.AltNames{},
+		Usages:       nil,
+	}
+
+	return c.configFileGenerator.GenTLSCert(c.Domain, regCertConfig)
+}
+
+func (c *localSingletonConfigurator) genBasicAuth() error {
+	//gen basic auth info: if not config, will skip.
+	if c.Auth.Username == "" || c.Auth.Password == "" {
+		return nil
+	}
+
+	return c.configFileGenerator.GenBasicAuth(c.Auth.Username, c.Auth.Password)
+}
+
+func (c *localSingletonConfigurator) initRegistry() error {
+	// bash init-registry.sh ${port} ${mountData} ${domain}
+	initRegistry := fmt.Sprintf("cd %s/scripts && bash init-registry.sh %s %s %s", c.rootfs, c.Port, c.DataDir, c.Domain)
+	if err := c.infraDriver.CmdAsync(c.DeployHost, initRegistry); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *localSingletonConfigurator) configHostsFile() error {
+	// add registry ip to "/etc/hosts"
+	var (
+		registryIP = c.DeployHost.String()
+		domain     = c.Domain
+		ips        = c.infraDriver.GetHostIPList()
+	)
+
+	hostsLine := registryIP + " " + domain
+	writeToHostsCmd := fmt.Sprintf("cat /etc/hosts |grep '%s' || echo '%s' >> /etc/hosts", hostsLine, hostsLine)
```

Comment:
take case: if  hostalias unchanged but ip changed, this command will fail.

In trident, we do as follow:
```go
func CommandSetHostAlias(hostName, ip string) string {
	return fmt.Sprintf(`if grep %s /etc/hosts;then sed -i "/%s/d" /etc/hosts; fi;echo "%s %s #%s" >>/etc/hosts`, hostName, hostName, ip, hostName, config.TridentSetAliasAnno)
}

func CommandUnSetHostAlias() string {
	return fmt.Sprintf(`sed -i "/%s/d" /etc/hosts`, config.TridentSetAliasAnno)
}

```